### PR TITLE
[MOBILE-253] Add support for determining subject type based off of grant

### PR DESCRIPTION
**Pull request by suwyn**
[MOBILE-253](https://grandrounds.atlassian.net/browse/MOBILE-253)

### CODE REVIEW

File: lib/grnds/sso/jwt/subject.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      HEADER = 'Authorization'.freeze
+      COOKIE = 'user_auth0_token'.freeze
+
+      module Subject
+        GRANT_TYPE_CLAIM = :'gty'.freeze
+        UID_CLAIM = :'https://grandrounds.com/claims/uid'.freeze
+        TYPES = {
+          'password': :user,
+          'client-credentials': :application
+        }.freeze
```
Comment (wyattEpp) [inline lib/grnds/sso/jwt/subject.rb]: Don't we need the 'implicit' grant type pointing to `:user` as well? Is that not the grant type given to tokens issued via implicit flow (Universal login page)?

I believe 'password' grant type refers specifically to resource owner services (trusted service supplies a user's creds to auth service directly to fetch a token on user's behalf).

File: lib/grnds/sso/version.rb [hunk 2]
```diff
@@ -1,5 +1,5 @@
 module Grnds
   module Sso
-    VERSION = '2.0'
+    VERSION = '2.1'.freeze
```
Comment (wyattEpp) [inline lib/grnds/sso/version.rb]: Can we adopt semantic versioning?

File: lib/grnds/sso/jwt/subject.rb (hunk 1 above)
Comment (wyattEpp) [reply lib/grnds/sso/jwt/subject.rb]: https://auth0.com/docs/applications/reference/grant-types-available

File: lib/grnds/sso/version.rb (hunk 2 above)
Comment (RoryO) [reply lib/grnds/sso/version.rb]: I'm okay with a straight integer number increase that most projects do now, like chrome, firefox, react, etc.

File: lib/grnds/sso/jwt/subject.rb (hunk 1 above)
Comment (suwyn) [reply lib/grnds/sso/jwt/subject.rb]: @wyattEpp I took a different approach that should handle it correctly.

File: spec/lib/grnds/sso/jwt/subject_spec.rb [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        expect(result.id).to eq(uid)
+      end
+    end
+
+    context 'without required subject claims' do
+      let(:token_type) { }
+      let(:uid) { 67890 }
+      let(:result) {
+        subject.call(request)
+      }
```
Comment (gauravs) [inline spec/lib/grnds/sso/jwt/subject_spec.rb]: Are there supposed to be tests here?

File: spec/lib/grnds/sso/jwt/subject_spec.rb [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+    context 'with a jwt set as a cookie' do
+      before do
+        allow(request).to receive(:cookies).and_return(jwt_cookies)
+      end
+
+      it_behaves_like 'it resolved a jwt'
+    end
+  end
+end
```
Comment (gauravs) [inline spec/lib/grnds/sso/jwt/subject_spec.rb]: :clap: great tests! :smiley:

File: spec/lib/grnds/sso/jwt/subject_spec.rb (hunk 3 above)
Comment (suwyn) [reply spec/lib/grnds/sso/jwt/subject_spec.rb]: @gauravs I removed this unused context.

**Pull request merged**